src/model/consent/consent.ts

```diff
@@ -0,0 +1,113 @@
+ http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, the Mojaloop files are distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ Contributors
+ --------------
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Raman Mangla <ramanmangla@google.com>
+ --------------
+ ******/
+
+import Knex from 'knex'
+
+/*
+ * Interface for Consent resource type
+ */
+export interface Consent {
+  id: string;
+  initiatorId: string;
+  participantId: string;
+  credentialId?: string;
+  credentialType?: string;
+  credentialStatus?: string;
+  credentialPayload?: string;
+  credentialChallenge?: string;
+  createdAt?: string;
+}
+
+/*
+ * Class to abstract Consent DB operations
+ */
+export class ConsentDB {
+  // Knex instance
+  private Db: Knex
+  // Nullable fields which don't allow for explicit null upsert
+  private static nullProtectedProps: Map<string, boolean> = new Map([
+    ['credentialId', true],
+    ['credentialType', true],
+    ['credentialStatus', true],
+    ['credentialPayload', true],
+    ['credentialChallenge', true]
+  ])
+
+  public constructor (dbInstance: Knex) {
+    this.Db = dbInstance
+  }
+
+  // Add initial Consent parameters
+  public register (consent: Consent): Promise<void> {
+    return this.Db<Consent>('Consent')
+      .insert(consent)
+  }
+
+  // Update Consent credential
+  // Only non-null Consent fields are updated
+  public async updateCredentials (consent: Consent): Promise<number> {
+    const validatedConsent = {}
+
+    for (const key in consent) {
+      // Only allow nullProtectedFields with non-null values for update
+      if (ConsentDB.nullProtectedProps.get(key) && Object(consent)[key]) {
+        Object(validatedConsent)[key] = Object(consent)[key]
+      }
+    }
+
+    const numUpdated: number = await this
+      .Db<Consent>('Consent')
+      .where({ id: consent.id })
+      .update(validatedConsent)
+
+    // Ensure that the caller knows that the resource does not exist
+    if (numUpdated === 0) {
+      throw Error('Error 404: Consent ID does not exist')
+    }
+
+    return numUpdated
+  }
+
+  // Retrieve Consent by ID
+  public retrieve (id: string): Promise<Consent[]> {
```

Comment:
Shouldn't we have a limit of 1 and always return a single matching consent or a null value? So this would be a Promise<Consent> returned and we'd include a limit() statement in the query?

Reply:
I agree that the return type of a Promise containing a single Consent/null would be better. I'll implement that. 

I didn't include a limit of 1 consent here because the DB cannot have replicated consentIDs in the first place. Should I still have a limit?

Reply:
Yeah, put the limit in, and just return the `Consent`. Keep in mind that those assumptions may change in the future (even though that's very unlikely), so putting them in the code explicitly here is a way to make sure our code breaks when our assumptions change.